views/js/controller/creator/helpers/changeTracker.js

```diff
@@ -0,0 +1,243 @@
+        leave: __('The test has unsaved changes, are you sure you want to leave?'),
+        exit: __('The test has unsaved changes, would you like to save it?')
+    };
+
+    /**
+     *
+     * @param {HTMLElement} container
+     * @param {testCreator} testCreator
+     * @param {String} [wrapperSelector]
+     * @returns {changeTracker}
+     */
+    function changeTrackerFactory(container, testCreator, wrapperSelector = 'body') {
+
+        // internal namespace for global registered events
+        const eventNS = `.ct-${uuid(8, 16)}`;
+
+        // keep the value of the test before changes
+        let originalTest;
+
+        // are we in the middle of the confirm process?
+        let asking = false;
+
+        /**
+         * @typedef {Object} changeTracker
+         */
+        const changeTracker = eventifier({
+            /**
+             * Initialized the changed state
+             * @returns {changeTracker}
+             */
+            init() {
+                originalTest = this.getSerializedTest();
+
+                return this;
+            },
+
+            /**
+             * Installs the change tracker, registers listeners
+             * @returns {changeTracker}
+             */
+            install() {
+                this.init();
+                asking = false;
+
+                // add a browser popup to prevent leaving the browser
+                $(window)
+                    .on(`beforeunload${eventNS}`, () => {
+                        if (!asking && this.hasChanged()) {
+                            return messages.leave;
+                        }
+                    })
+                    // since we don't know how to prevent history based events, we just stop the handling
+                    .on('popstate', this.uninstall);
+
+                // every click outside the authoring
+                $(wrapperSelector)
+                    .on(`click${eventNS}`, e => {
+                        if (!$.contains(container, e.target) && this.hasChanged()) {
+                            e.stopImmediatePropagation();
+                            e.preventDefault();
+
+                            this.confirmBefore('exit')
+                                .then(whatToDo => {
+                                    this.ifWantSave(whatToDo);
+                                    this.uninstall();
+                                    e.target.click();
+                                })
+                                //do nothing but prevent uncaught error
+                                .catch(() => {});
+                        }
+                    });
+
+                testCreator
+                    .on(`ready${eventNS} saved${eventNS}`, this.init)
+                    .before(`creatorclose${eventNS}`, () => this.confirmBefore('exit').then(whatToDo => {
+                        this.ifWantSave(whatToDo);
+                        window.history.back();
+                        this.uninstall;
```

Comment:
Don't know if I missed something, but should not be `this.uninstall()` instead of `this.uninstall`?

Reply:
Yes :) Done